Fix edge cases in IPInt tail calls

danlliu opened the pull request:
#### 602c7b2295a8dd9f3bedfa9f23d09579f682c11d
<pre>
Fix edge cases in IPInt tail calls
<a href="https://bugs.webkit.org/show_bug.cgi?id=284785">https://bugs.webkit.org/show_bug.cgi?id=284785</a>
<a href="https://rdar.apple.com/141575550">rdar://141575550</a>

Reviewed by Yusuke Suzuki.

IPInt&apos;s tail calls previously failed on changes in the stack frame layout, since the stack pointer
was not correctly updated prior to the tail call. This patch resolves that issue, as well as a few
others related to tail calls. IPInt is able to pass the `wasm/stress/tail-call.js` stress test
consistently now.

* JSTests/wasm/ipint-tests/ipint-test-tail-call-doubleargs.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.func.param.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.result.i32.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.local.9.local.10.local.11.local.12.local.13.local.14.local.15.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.func.main.export.string_appeared_here.param.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.result.i32.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.local.9.local.10.local.11.local.8.local.9.local.10.local.11.return_call.func.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-fpargs.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.func.param.f32.f32.f32.f32.i32.i32.i32.i32.i32.i32.i32.i32.f32.f32.f32.f32.f32.f32.f32.f32.result.f32.local.0.local.1.local.2.local.3.local.12.local.13.local.14.local.15.local.16.local.17.local.18.local.19.f32.add.f32.add.f32.add.f32.add.f32.add.f32.add.f32.add.f32.add.f32.add.f32.add.f32.add.func.main.export.string_appeared_here.param.f32.f32.f32.f32.i32.i32.i32.i32.i32.i32.i32.i32.f32.f32.f32.f32.f32.f32.f32.f32.result.f32.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.local.9.local.10.local.11.local.12.local.13.local.14.local.15.local.16.local.17.local.18.local.19.return_call.func.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-indirect-doubleargs.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.type.sig.func.param.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.result.i32.table.table.export.string_appeared_here.1.funcref.elem.i32.const.0.func.func.func.param.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.result.i32.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.local.9.local.10.local.11.local.12.local.13.local.14.local.15.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.func.main.export.string_appeared_here.param.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.result.i32.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.local.9.local.10.local.11.local.8.local.9.local.10.local.11.i32.const.0.return_call_indirect.table.type.sig.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-manyargs.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.func.param.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.result.i32.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.local.9.local.10.local.11.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.func.main.export.string_appeared_here.param.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.result.i32.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.local.9.local.10.local.11.return_call.func.func.func2.param.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.result.i32.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.local.9.local.10.local.11.local.12.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.func.main2.export.string_appeared_here.param.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.result.i32.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.local.9.local.10.local.11.local.12.return_call.func2.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-manyret.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.helper.result.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.const.0.i32.const.1.i32.const.2.i32.const.3.i32.const.4.i32.const.5.i32.const.6.i32.const.7.i32.const.8.i32.const.9.i32.const.10.i32.const.11.func.main.export.string_appeared_here.result.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.return_call.helper.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-stress.js: Added.
(i32.const.0.func.0.1.2.3.4.5.6.7.8.9.10.11.12.13.14.15.16.17.18.19.20.21.22.23.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-stress2.js: Added.
(i64.const.30.call.6.export.string_appeared_here.func.0.export.string_appeared_here.func.1.export.string_appeared_here.func.2.export.string_appeared_here.func.3.export.string_appeared_here.func.4.export.string_appeared_here.func.5.export.string_appeared_here.func.6.export.string_appeared_here.func.7.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-stress3.js: Added.
(type.1.param.i32.i64.f32.i64.f32.f64.f32.f64.i32.i32.i32.f64.f64.i64.f64.i32.i64.i64.f32.f64.f32.i32.i64.i64.f64.f64.result.i64.i64.f64.i32.i32.f64.i64.f64.f64.i32.i64.i32.f64.f32.f64.i64.local.2.local.1.local.0.local.8.local.5.local.7.local.9.return_call.0.export.string_appeared_here.func.0.export.string_appeared_here.func.1.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-stress4.js: Added.
(24.funcref.export.string_appeared_here.func.0.export.string_appeared_here.func.1.export.string_appeared_here.func.2.export.string_appeared_here.func.3.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-stress5.js: Added.
(type.1.result.f64.i32.f32.f32.const.8.f64.const.9.i64.const.10.i32.const.11.f64.const.12.i64.const.13.f32.const.14.f64.const.15.i32.const.16.f32.const.17.i32.const.18.f64.const.19.i32.const.20.f32.const.21.f64.const.22.i32.const.23.f64.const.24.i32.const.25.f32.const.26.i64.const.27.f64.const.28.i32.const.29.i32.const.30.f64.const.31.f32.const.32.f32.const.33.i64.const.34.f32.const.35.i32.const.36.f32.const.37.i32.const.38.f64.const.39.f32.const.40.i32.const.41.i32.const.42.i64.const.43.i64.const.44.i64.const.45.i64.const.46.return_call.0.export.string_appeared_here.func.0.export.string_appeared_here.func.1.async test):
* Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
* Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
* Source/JavaScriptCore/llint/LLIntSlowPaths.h:
* Source/JavaScriptCore/wasm/WasmCalleeGroup.cpp:
(JSC::Wasm::CalleeGroup::CalleeGroup):
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.cpp:
(JSC::Wasm::FunctionIPIntMetadataGenerator::setTailCall):
(JSC::Wasm::FunctionIPIntMetadataGenerator::addReturnData):
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.h:
(JSC::Wasm::FunctionIPIntMetadataGenerator::setTailCallClobbersInstance):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::assertAboutStackSize):
(JSC::Wasm::IPIntGenerator::didParseOpcode):
(JSC::Wasm::IPIntGenerator::addBlock):
(JSC::Wasm::IPIntGenerator::addLoop):
(JSC::Wasm::IPIntGenerator::addTryTable):
(JSC::Wasm::IPIntGenerator::addCatchToUnreachable):
(JSC::Wasm::IPIntGenerator::addCatchAllToUnreachable):
(JSC::Wasm::IPIntGenerator::addSwitch):
(JSC::Wasm::IPIntGenerator::addCallCommonData):
(JSC::Wasm::IPIntGenerator::addTailCallCommonData):
(JSC::Wasm::IPIntGenerator::addCall):
(JSC::Wasm::IPIntGenerator::addCallIndirect):
(JSC::Wasm::IPIntGenerator::addCallRef):

Canonical link: <a href="https://commits.webkit.org/288056@main">https://commits.webkit.org/288056@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a79d88ea7a1c9a107a01eb2e1448e10b7ce82a76

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81757 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1283 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35711 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/86304 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32753 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1316 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9103 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/63783 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21499 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84827 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/957 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74398 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44069 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/857 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28573 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31206 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/74737 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72235 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29176 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87740 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/80811 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8997 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6389 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72116 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9182 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70218 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71346 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17775 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15437 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14354 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/103223 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8948 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14480 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25063 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8789 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12312 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10597 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
```diff
@@ -2764,6 +2778,10 @@ void IPIntGenerator::addTailCallCommonData(const FunctionSignature& signature)
+
+    uint32_t numStackValues = WTF::roundUpToMultipleOf(stackAlignmentRegisters(), wasmCallingConvention().numberOfStackValues(signature));
+    ASSERT(!(numStackValues % 2));
+    m_metadata->appendMetadata(numStackValues);
```
Don't we need to update the max frame size by this amount since we copy the frame right? Would be worth adding a test for this.

kmiller68 commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -5756,12 +5772,46 @@ end
+    #  argument
+    #  return val
+    #  return val
+    #  <caller frame>
```
In the rest of JSC we have the caller's frame textually above the callee's frame as the stack grows down in memory. I think we should be consistent here.

kmiller68 commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -5756,12 +5772,46 @@ end
+    #  argument
+    #  return val
+    #  return val                  <- sc2
+    #  <caller frame>
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -5778,6 +5828,25 @@ end
+    #  argument
+    #  return val
+    #  return val                  <- sc2
+    #  <caller frame>
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -5955,12 +6028,15 @@ _wasm_ipint_call_return_location_wide32:
-    move sp, mintRetSrc
+    loadi [MC], sc1
+    advanceMC(4)
+    leap [sp, mintRetSrc], mintRetSrc
```
Isn't this just addp?

danlliu replied:
It should be an `addp`, but the LLInt doesn't like using `addp` with `sp` on ARM, so I'm replacing it with an equivalent `leap` to keep it as one instruction (and not increase register pressure).

kmiller68 commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -6112,11 +6188,63 @@ end
+    #  argument
+    #  return val
+    #  return val                  <- sc2
+    #  <caller frame>
```
Ditto.

kmiller68 commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -6125,11 +6253,34 @@ end
+    #  ...
+    #  return val
+    #  return val
+    #  <caller frame>
```
Ditto.

Constellation reviewed (approved):
Looks good. Let's resolve Keith's comment too.

danlliu committed:
Fix edge cases in IPInt tail calls
https://bugs.webkit.org/show_bug.cgi?id=284785
rdar://141575550

Reviewed by Yusuke Suzuki.

IPInt's tail calls previously failed on changes in the stack frame layout, since the stack pointer
was not correctly updated prior to the tail call. This patch resolves that issue, as well as a few
others related to tail calls. IPInt is able to pass the `wasm/stress/tail-call.js` stress test
consistently now.

* JSTests/wasm/ipint-tests/ipint-test-tail-call-doubleargs.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.func.param.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.result.i32.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.local.9.local.10.local.11.local.12.local.13.local.14.local.15.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.func.main.export.string_appeared_here.param.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.result.i32.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.local.9.local.10.local.11.local.8.local.9.local.10.local.11.return_call.func.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-fpargs.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.func.param.f32.f32.f32.f32.i32.i32.i32.i32.i32.i32.i32.i32.f32.f32.f32.f32.f32.f32.f32.f32.result.f32.local.0.local.1.local.2.local.3.local.12.local.13.local.14.local.15.local.16.local.17.local.18.local.19.f32.add.f32.add.f32.add.f32.add.f32.add.f32.add.f32.add.f32.add.f32.add.f32.add.f32.add.func.main.export.string_appeared_here.param.f32.f32.f32.f32.i32.i32.i32.i32.i32.i32.i32.i32.f32.f32.f32.f32.f32.f32.f32.f32.result.f32.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.local.9.local.10.local.11.local.12.local.13.local.14.local.15.local.16.local.17.local.18.local.19.return_call.func.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-indirect-doubleargs.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.type.sig.func.param.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.result.i32.table.table.export.string_appeared_here.1.funcref.elem.i32.const.0.func.func.func.param.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.result.i32.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.local.9.local.10.local.11.local.12.local.13.local.14.local.15.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.func.main.export.string_appeared_here.param.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.result.i32.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.local.9.local.10.local.11.local.8.local.9.local.10.local.11.i32.const.0.return_call_indirect.table.type.sig.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-manyargs.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.func.param.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.result.i32.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.local.9.local.10.local.11.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.func.main.export.string_appeared_here.param.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.result.i32.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.local.9.local.10.local.11.return_call.func.func.func2.param.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.result.i32.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.local.9.local.10.local.11.local.12.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.i32.add.func.main2.export.string_appeared_here.param.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.result.i32.local.0.local.1.local.2.local.3.local.4.local.5.local.6.local.7.local.8.local.9.local.10.local.11.local.12.return_call.func2.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-manyret.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.helper.result.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.const.0.i32.const.1.i32.const.2.i32.const.3.i32.const.4.i32.const.5.i32.const.6.i32.const.7.i32.const.8.i32.const.9.i32.const.10.i32.const.11.func.main.export.string_appeared_here.result.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.i32.return_call.helper.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-stress.js: Added.
(i32.const.0.func.0.1.2.3.4.5.6.7.8.9.10.11.12.13.14.15.16.17.18.19.20.21.22.23.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-stress2.js: Added.
(i64.const.30.call.6.export.string_appeared_here.func.0.export.string_appeared_here.func.1.export.string_appeared_here.func.2.export.string_appeared_here.func.3.export.string_appeared_here.func.4.export.string_appeared_here.func.5.export.string_appeared_here.func.6.export.string_appeared_here.func.7.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-stress3.js: Added.
(type.1.param.i32.i64.f32.i64.f32.f64.f32.f64.i32.i32.i32.f64.f64.i64.f64.i32.i64.i64.f32.f64.f32.i32.i64.i64.f64.f64.result.i64.i64.f64.i32.i32.f64.i64.f64.f64.i32.i64.i32.f64.f32.f64.i64.local.2.local.1.local.0.local.8.local.5.local.7.local.9.return_call.0.export.string_appeared_here.func.0.export.string_appeared_here.func.1.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-stress4.js: Added.
(24.funcref.export.string_appeared_here.func.0.export.string_appeared_here.func.1.export.string_appeared_here.func.2.export.string_appeared_here.func.3.async test):
* JSTests/wasm/ipint-tests/ipint-test-tail-call-stress5.js: Added.
(type.1.result.f64.i32.f32.f32.const.8.f64.const.9.i64.const.10.i32.const.11.f64.const.12.i64.const.13.f32.const.14.f64.const.15.i32.const.16.f32.const.17.i32.const.18.f64.const.19.i32.const.20.f32.const.21.f64.const.22.i32.const.23.f64.const.24.i32.const.25.f32.const.26.i64.const.27.f64.const.28.i32.const.29.i32.const.30.f64.const.31.f32.const.32.f32.const.33.i64.const.34.f32.const.35.i32.const.36.f32.const.37.i32.const.38.f64.const.39.f32.const.40.i32.const.41.i32.const.42.i64.const.43.i64.const.44.i64.const.45.i64.const.46.return_call.0.export.string_appeared_here.func.0.export.string_appeared_here.func.1.async test):
* Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
* Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
* Source/JavaScriptCore/llint/LLIntSlowPaths.h:
* Source/JavaScriptCore/wasm/WasmCalleeGroup.cpp:
(JSC::Wasm::CalleeGroup::CalleeGroup):
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.cpp:
(JSC::Wasm::FunctionIPIntMetadataGenerator::setTailCall):
(JSC::Wasm::FunctionIPIntMetadataGenerator::addReturnData):
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.h:
(JSC::Wasm::FunctionIPIntMetadataGenerator::setTailCallClobbersInstance):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::assertAboutStackSize):
(JSC::Wasm::IPIntGenerator::didParseOpcode):
(JSC::Wasm::IPIntGenerator::addBlock):
(JSC::Wasm::IPIntGenerator::addLoop):
(JSC::Wasm::IPIntGenerator::addTryTable):
(JSC::Wasm::IPIntGenerator::addCatchToUnreachable):
(JSC::Wasm::IPIntGenerator::addCatchAllToUnreachable):
(JSC::Wasm::IPIntGenerator::addSwitch):
(JSC::Wasm::IPIntGenerator::addCallCommonData):
(JSC::Wasm::IPIntGenerator::addTailCallCommonData):
(JSC::Wasm::IPIntGenerator::addCall):
(JSC::Wasm::IPIntGenerator::addCallIndirect):
(JSC::Wasm::IPIntGenerator::addCallRef):

Canonical link: https://commits.webkit.org/288056@main
